src/learning/gmm.rs

```diff
@@ -209,174 +212,404 @@ impl GaussianMixtureModel {
+        let precisions: &Vec<Matrix<f64>> = self.precisions.as_ref().unwrap();
+        let ref model_means = self.model_means.as_ref().unwrap();
+        // The log of the determinant for each precision matrix
+        let log_det = precisions.iter().map(|m| m.det().ln());
+        // println!("log_det: {:?}", log_det);
+
+        let mut log_prob = Matrix::zeros(inputs.rows(), self.comp_count);
+        for k in 0..self.comp_count {
+            let prec = &precisions[k];
+            // y is a matrix of shape [n_samples, n_features]
+            let mut y = inputs * prec;
+            // println!("y: \n{:.4}", &y.select_rows(&[0, 1, 2, 3, 4, 5, 6]));
+            // Matrix of shape [1, n_features]
+            let z: Matrix<f64> = model_means.select_rows(&[k]) * prec;
+            // println!("z: \n{:.4}", &z);
+            
+            // Subtract the mean of each column from the matrix y
+            for col in 0..y.cols() {
+                for row in 0..y.rows() {
+                    y[[row, col]] -= z[[0, col]];
                 }
-                cov_mat *= reg_value;
-                if let CovOption::Regularized(eps) = self.cov_option {
-                    cov_mat += Matrix::<f64>::identity(cov_mat.cols()) * eps;
+            }
+            // println!("y: \n{:.4}", &y.select_rows(&[0, 1, 2, 3, 4, 5, 6]));
+
+            for (i, row) in y.iter_rows().enumerate() {
+                let sum_of_squares = row.iter().map(|v| v.powi(2)).sum();
+                log_prob[[i, k]] = sum_of_squares;
+            }
+        }
+
+        log_prob = (log_prob + inputs.cols() as f64 * (2.0 * PI).ln()) * -0.5;
+        for (row, det) in log_prob.iter_rows_mut().zip(log_det) {
+            for v in row {
+                *v += det;
+            }
+        }
+        log_prob
+    }
+
+    fn estimate_log_prob_resp(&self, inputs: &Matrix<f64>) -> (Vector<f64>, Matrix<f64>) {
+        let mut weighted_log_prob: Matrix<f64> = self.estimate_weighted_log_prob(inputs);
+        // length of n_samples
+        let log_prob_norm: Vector<f64> = 
+            Vector::new(weighted_log_prob.iter_rows().map(|row: &[f64]| {
+                let a: f64 = row.iter().map(|v| v.exp()).sum();
+                a.ln()
+            }).collect::<Vec<f64>>());
+
+        for row in 0..log_prob_norm.size() {
+            for col in 0..weighted_log_prob.cols() {
+                weighted_log_prob[[row, col]] -= log_prob_norm[row];
+            }
+        }
+
+        (log_prob_norm, weighted_log_prob)
+    }
+
+    fn update_gaussian_parameters(&mut self, inputs: &Matrix<f64>, resp: Matrix<f64>) {
+        self.mix_weights = resp.iter_rows()
+            .fold(Vector::new(vec![0.0; resp.cols()]), |mut acc, row| {
+                for (a, r) in acc.iter_mut().zip(row.iter()) {
+                    *a += *r;
                 }
-                Ok(cov_mat)
+                acc
+            })
+            + 10.0 * EPSILON;
```

Comment:
This seems like black magic to me - I see it in the sklearn implementation but we should try to find some sources for it ourselves. Is it purely for numeric stability?

And a (very) minor point, I think it is better to `use f64` and here call `f64::EPSILON`. I think this makes the constant clearer, otherwise someone may think we have defined `EPSILON`.